src/server/dfly_main.cc

```diff
@@ -456,103 +456,129 @@ bool ShouldUseEpollAPI(const base::sys::KernelVersion& kver) {
-   * are flawed, however together do cover a
-   * good portion of cases.
-   *
-   * 1. Checking '/.dockerenv': very simple, however
-   * only works for Docker, and this file may be moved
-   * in the future.
-   * 2. Checking '/sys/fs/cgroup/memory.max': This
-   * directory -cannot- be edited (not even using sudoedit),
-   * so users are not able to add files to there. The file
-   * 'memory.max' should contain a memory limit for processes.
-   *
-   * We also use this file to find out how much memory we can use.
-   * However, on LXC this is placed in a different directory:
-   * 3. Checking '/sys/fs/cgroup/memory/memory.limit_in_bytes':
-   * Same idea.
-   */
-
   using io::Exists;
 
-  auto cgroup_res = GetCGroupPath();
-  if (!cgroup_res) {
-    VLOG(1) << "Failed to get cgroup path, error: " << cgroup_res.error().message();
+  string mem_path, cpu_path;
+  auto err = GetCGroupPath(&mem_path, &cpu_path);
+
+  if (mem_path.empty() || cpu_path.empty()) {
+    VLOG(1) << "Failed to get cgroup path, error: " << err;
     return;
   }
 
-  const string cgroup = "/sys/fs/cgroup/" + *cgroup_res;
-
-  if (!Exists(cgroup + "/memory.max"))
-    return;
+  auto original_memory = mdata->mem_total;
 
   /* Update memory limits */
-  auto mlimit = io::ReadFileToString(cgroup + "/memory/memory.limit_in_bytes");
+  auto mlimit = io::ReadFileToString(mem_path + "/memory.limit_in_bytes");
   DVLOG(1) << "memory/memory.limit_in_bytes: " << mlimit.value_or("N/A");
 
   if (mlimit && !absl::StartsWith(*mlimit, "max")) {
     CHECK(absl::SimpleAtoi(*mlimit, &mdata->mem_total));
   }
 
-  auto mmax = io::ReadFileToString(cgroup + "/memory.max");
+  auto mmax = io::ReadFileToString(mem_path + "/memory.max");
   DVLOG(1) << "memory.max: " << mmax.value_or("N/A");
 
   if (mmax && !absl::StartsWith(*mmax, "max")) {
     CHECK(absl::SimpleAtoi(*mmax, &mdata->mem_total));
   }
 
   mdata->mem_avail = mdata->mem_total;
-  auto mhigh = io::ReadFileToString(cgroup + "/memory.high");
+  auto mhigh = io::ReadFileToString(mem_path + "/memory.high");
 
   if (mhigh && !absl::StartsWith(*mhigh, "max")) {
     CHECK(absl::SimpleAtoi(*mhigh, &mdata->mem_avail));
   }
 
+  if (numeric_limits<size_t>::max() == mdata->mem_total)
+    mdata->mem_avail = original_memory;
+
   /* Update thread limits */
 
-  if (auto cpu = ReadFileToString(cgroup + "/cpu.max"); cpu.has_value()) {
+  double count = 0, timeshare = 1;
+
+  if (auto cpu = ReadFileToString(cpu_path + "/cpu.max"); cpu.has_value()) {
     vector<string_view> res = absl::StrSplit(cpu.value(), ' ');
 
     CHECK_EQ(res.size(), 2u);
 
     if (res[0] == "max")
       *max_threads = 0u;
     else {
-      double count, timeshare;
       CHECK(absl::SimpleAtod(res[0], &count));
       CHECK(absl::SimpleAtod(res[1], &timeshare));
 
       *max_threads = static_cast<size_t>(ceil(count / timeshare));
     }
-  } else
+  } else if (auto quota = ReadFileToString(cpu_path + "/cpu.cfs_quota_us"); quota.has_value()) {
```

Comment:
how have you tested this  @talbii ?

can you please add into PR/commit description how to reproduce cgroups v1 behavior?

Reply:
On it. The squashed commit will contain details for both v1 and v2.